content/Training/developers/05-guardrails-evals-observability.mdx

```diff
@@ -0,0 +1,1114 @@
+---
+title: "Module 5: Guardrails, Evals & Observability"
+description: Making agents reliable, safe, and production-ready
+---
+
+You've built agents that can think, remember, and collaborate. Now it's time to make them production-ready.
+
+## The Reality of Production Agents
+
+When [Salesforce deployed their Agentforce AI agents](https://www.hr-brew.com/stories/2025/03/04/salesforce-ai-agents-reskilling), they discovered that success required more than just technology - it demanded a comprehensive reskilling strategy for their 72,000+ employees. As their EVP of talent growth noted: "This rise of digital labor powered by AI agents is truly reshaping the way our businesses operate." The gap between demo and production? Proper guardrails, systematic evaluation, and comprehensive observability.
+
+According to [NIST's AI Risk Management framework](https://www.nist.gov/itl/ai-risk-management-framework), the primary operational risks in AI systems include:
+- **Hallucination**: Agents generating plausible but incorrect information
+- **Prompt Injection**: Adversarial inputs manipulating agent behavior
+- **Resource Consumption**: Uncontrolled usage leading to excessive costs
+- **Compliance Drift**: Agents violating domain-specific regulations
+
+## The Three Pillars of Production Agents
+
+<Callout type="info">
+Agentuity provides built-in OpenTelemetry integration with automatic instrumentation through Traceloop SDK, giving you comprehensive observability out of the box.
+</Callout>
+
+### 1. Guardrails: Setting Boundaries
+
+Guardrails prevent agents from harmful actions while preserving autonomy:
+
+- **Input Validation**: Schema enforcement, content filtering, size limits
+- **Rate Limiting**: Prevent abuse and control costs per user/session  
+- **Security**: Prompt injection defense ([WASP](https://arxiv.org/abs/2407.01593)), tool permissions
+- **Domain Rules**: Compliance checks, output validation, custom constraints
+
+### 2. Evaluation: Measuring Success
+
+Systematic evaluation is critical for non-deterministic agents:
+
+- **Industry Benchmarks**: [SWE-Bench](https://www.swebench.com/) (success rate on software engineering tasks)
+- **Automated Testing**: Unit tests, integration tests, golden datasets
+- **Production Metrics**: Success rates, latency, cost per request
+- **A/B Testing**: Shadow deployments, gradual rollouts
+
+### 3. Observability: Seeing Everything
+
+Agentuity provides automatic OpenTelemetry integration with Traceloop SDK:
+
+- **What's Tracked**: LLM calls, tool invocations, storage operations, API calls
+- **Three Pillars**: Logs (events), Metrics (measurements), Traces (request flow)
```

Comment:
"Three pillars" here and then there's a top level heading of three pillars, which is different.  Kinda confusing.